libbpf-tools/tcpdrop.c

```diff
@@ -0,0 +1,489 @@
+/* SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause) */
+
+/*
+ * tcpdrop: Trace TCP kernel-dropped packets/segments using eBPF
+ * Copyright (c) 2025 Zi Li <zi.li@linux.dev>
+ *
+ * Based on tcptop(8) from BCC by Brendan Gregg.
+ *
+ * 01-May-2025   Lance Yang    Created initial demo.
+ * 10-Jun-2025   Zi Li         Created this.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+#include <unistd.h>
+#include <signal.h>
+#include <time.h>
+#include <sys/stat.h>
+
+#include <arpa/inet.h>
+#include <bpf/bpf.h>
+#include <bpf/libbpf.h>
+#include <linux/types.h>
+#include <argp.h>
+
+#include "tcpdrop.skel.h"
+
+#define warn(...) fprintf(stderr, __VA_ARGS__)
+
+#define ARRAY_SIZE(x) (sizeof(x) / sizeof((x)[0]))
+#define TASK_COMM_LEN 16
+#define REASON_MAX_LEN 64
+#define MAX_DROP_REASONS 128
+#define MAX_STACK_DEPTH 15
+#define MAX_SYMBOL_NAME_LEN 128
+#define MAX_SYMBOLS 200000
+#define NANOSECONDS_IN_SECOND 1000000000
+
+static volatile sig_atomic_t exiting = 0;
+static struct tcpdrop_bpf *skel = NULL;
+static char drop_reasons[MAX_DROP_REASONS][REASON_MAX_LEN];
+static int drop_reason_max = 0;
+static bool drop_reason_inited = false;
+
+const char *argp_program_version = "tcpdrop 0.1";
+const char *argp_program_bug_address =
+	"https://github.com/iovisor/bcc/tree/master/libbpf-tools";
+const char argp_program_doc[] =
+"Trace TCP kernel-dropped packets/segments using eBPF.\n"
+"\n"
+"USAGE: tcpdrop [-h] [-4] [-6] [--netns-id ID] [--pid-netns PID]\n"
+"\n"
+"EXAMPLES:\n"
+"    tcpdrop            # trace all TCP drops\n"
+"    tcpdrop -4         # trace IPv4 family only\n"
+"    tcpdrop -6         # trace IPv6 family only\n"
+"    tcpdrop --netns-id 12345  # filter by network namespace ID\n"
+"    tcpdrop --pid-netns 123   # filter by network namespace of PID\n";
+
+static const struct argp_option opts[] = {
+    { "ipv4", '4', NULL, 0, "trace IPv4 family only", 0 },
+    { "ipv6", '6', NULL, 0, "trace IPv6 family only", 0 },
+    { "netns-id", 'n', "ID", 0, "filter by network namespace ID", 0 },
+    { "pid-netns", 'p', "PID", 0, "filter by network namespace of PID", 0 },
+    { NULL, 'h', NULL, OPTION_HIDDEN, "show the full help", 0 },
+    {},
+};
+
+struct args {
+    bool ipv4_only;
+    bool ipv6_only;
+    __u32 netns_id;
+    __u32 pid_netns;
+};
+static struct args arguments = {0};
+
+static const char *tcp_states[] = {
+    [0] = "CLOSED",      [1] = "LISTEN",     [2] = "SYN_SENT",   [3] = "SYN_RECV",
+    [4] = "ESTABLISHED", [5] = "FIN_WAIT1",  [6] = "FIN_WAIT2",  [7] = "CLOSE_WAIT",
+    [8] = "CLOSING",     [9] = "LAST_ACK",   [10] = "TIME_WAIT", [11] = "DELETE_TCB",
+};
+
+struct event {
```

Comment:
Move this struct to a .h file instead so that you don't have to define it twice.